webapp/src/js/bootstrapper/initial-replication.js

```diff
@@ -0,0 +1,188 @@
+    $('.bootstrap-layer .error').html(content);
+    $('#btn-continue').click(() => resolve());
+    $('#btn-abort').click(() => {
+      document.cookie = 'login=force;path=/';
+      window.location.reload(false);
+    });
+  });
+};
+
+const getMissingDocIdsRevsPairs = async (localDb, remoteDocIdsRevs) => {
+  const localDocs = await getLocalDocList(localDb);
+  return remoteDocIdsRevs.filter(({ id, rev }) => !localDocs[id] || localDocs[id] !== rev);
+};
+
+const getDownloadList = async (localDb) => {
+  const response = await utils.fetchJSON('/api/v1/initial-replication/get-ids');
+
+  docIdsRevs = await getMissingDocIdsRevsPairs(localDb, response.doc_ids_revs);
+  lastSeq = response.last_seq;
+  remoteDocCount = response.doc_ids_revs.length;
+
+  if (response.warn) {
+    await displayTooManyDocsWarning(response);
+  }
+};
+
+const getLocalDocList = async (localDb) => {
+  const response = await localDb.allDocs();
+
+  const localDocMap = {};
+  response.rows.forEach(row => localDocMap[row.id] = row.value && row.value.rev);
+  return localDocMap;
+};
+
+const getDocsBatch = async (remoteDb, localDb) => {
+  const batch = docIdsRevs.splice(0, BATCH_SIZE);
+
+  if (!batch.length) {
+    return;
+  }
+
+  const res = await remoteDb.bulkGet({ docs: batch, attachments: true });
+  const docs = res.results
+    .map(result => result.docs && result.docs[0] && result.docs[0].ok)
+    .filter(doc => doc);
+  await localDb.bulkDocs(docs, { new_edits: false });
+};
+
+const downloadDocs = async (remoteDb, localDb) => {
+  setUiStatus('FETCH_INFO', { count: remoteDocCount - docIdsRevs.length, total: remoteDocCount });
+  do {
+    await getDocsBatch(remoteDb, localDb);
+    setUiStatus('FETCH_INFO', { count: remoteDocCount - docIdsRevs.length, total: remoteDocCount });
+  } while (docIdsRevs.length > 0);
+};
+
+const writeCheckpointers = async (remoteDb, localDb) => {
+  const replicateFromPromise = localDb.replicate.from(remoteDb, {
+    live: false,
+    retry: false,
+    heartbeat: 10000,
+    timeout: 1000 * 60 * 10, // try for ten minutes then give up,
+    query_params: { initial_replication: true },
+    since: lastSeq,
+  });
+
+  replicateFromPromise.on('change', () => replicateFromPromise.cancel());
+  await replicateFromPromise;
+
+  const localInfo = await localDb.info();
+  await localDb.replicate.to(remoteDb, {
+    since: localInfo.update_seq,
+  });
+};
+
+const replicate = async (remoteDb, localDb) => {
+  setUiStatus('LOAD_APP');
+
+  await startInitialReplication(localDb);
+
+  setUiStatus('POLL_REPLICATION');
+  await getDownloadList(localDb);
+  await downloadDocs(remoteDb, localDb);
+  await writeCheckpointers(remoteDb, localDb);
+
+  await completeInitialReplication(localDb);
+};
+
+const getReplicationLog = async (localDb) => {
+  try {
+    return await localDb.get(INITIAL_REPLICATION_LOG);
+  } catch (err) {
+    return null;
+  }
+};
+const isReplicationNeeded = async (localDb, userCtx) => {
```

Comment:
so if initial replication fails for whatever reason: when the user retries, they will get the slow/expensive replication experience if these few docs were successful...

why don't we put them into this 10x faster experience until initial replication fully completes?

Reply:
I think that's what the `replicationLog.complete` check does on line 178, so you have to have the required docs AND have a complete replication log before you go to slow mode.

Reply:
@garethbowen is correct. I create a flag doc before starting the quick initial replication. The flag is created without a `complete` property. The quick initial replication keep restarting on every refresh until it completes.